shared/domains/chelonia/chelonia.js

```diff
@@ -818,9 +820,28 @@ export default (sbp('sbp/selectors/register', {
   //       the events one-by-one instead of converting to giant json object?
   //       however, note if we do that they would be processed in reverse...
   'chelonia/out/eventsAfter': async function (contractID: string, since: string) {
-    const events = await fetch(`${this.config.connectionURL}/eventsAfter/${contractID}/${since}`, { signal: this.abortController.signal })
+    let events = await fetch(`${this.config.connectionURL}/eventsAfter/${contractID}/${since}`, { signal: this.abortController.signal })
       .then(handleFetchResult('json'))
     if (Array.isArray(events)) {
+      // Sanity check
+      if (GIMessage.deserialize(b64ToStr(events[events.length - 1])).hash() !== since) {
+        throw new Error('hash() !== since')
+      }
+      // Maybe we didn't receive all the requested events because of eventsAfter's limit.
+      if (MAX_EVENTS_AFTER && (events.length === MAX_EVENTS_AFTER)) {
+        while (true) {
+          const intermediateEventHash = GIMessage.deserialize(b64ToStr(events[0])).hash()
+          const nextEvents = await fetch(`${this.config.connectionURL}/eventsAfter/${contractID}/${intermediateEventHash}`)
+            .then(handleFetchResult('json'))
+          // Break if we didn't receive any event we didn't have yet.
+          // Note: nextEvents usually ends with an intermediate event we already have.
+          if (!Array.isArray(nextEvents) || nextEvents.length < 2) break
+          nextEvents.pop()
+          events = [...nextEvents, ...events]
+          // Only continue if we hit the limit again.
+          if (nextEvents.length !== MAX_EVENTS_AFTER - 1) break
+        }
+      }
```

Comment:
Reminder: call to `.reverse()` below can be removed once they're no longer reversed